crates/full-node/sov-sequencer/src/preferred/preferred_blob_sender.rs

```
@@ -14,65 +14,52 @@ use tokio::task::JoinHandle;
 use tokio::time::Duration;
 use tracing::debug;
 
-use super::db::{PreferredSequencerReadBatch, PreferredSequencerReadBlob};
+use super::db::{ReadBatch, ReadBlob};
 use crate::{common::TxStatusBlobSenderHooks, TxStatusManager};
 
 /// Wrapper around [`BlobSender`] with preferred blob -specific logic.
 pub struct PreferredBlobSender<Da: DaService> {
-    inner: Option<BlobSender<Da, TxStatusBlobSenderHooks<Da::Spec>, LedgerDb>>,
+    inner: BlobSender<Da, TxStatusBlobSenderHooks<Da::Spec>, LedgerDb>,
     nb_of_concurrent_blob_submissions: Arc<AtomicUsize>,
 }
 
 impl<Da: DaService> PreferredBlobSender<Da> {
     pub(crate) async fn new(
         da: Da,
         ledger_db: LedgerDb,
-        all_completed_blobs: Vec<PreferredSequencerReadBlob>,
+        all_completed_blobs: Vec<ReadBlob>,
         storage_path: Box<Path>,
         tx_status_manager: TxStatusManager<Da::Spec>,
         shutdown_sender: watch::Sender<()>,
         blob_processing_timeout: Duration,
         blobs_sender_channel: broadcast::Sender<BlobExecutionStatus<Da::Spec>>,
-        is_replica: bool,
-    ) -> anyhow::Result<(Self, Option<JoinHandle<()>>)> {
+    ) -> anyhow::Result<(Self, JoinHandle<()>)> {
         let nb_of_concurrent_blob_submissions = Arc::new(AtomicUsize::new(0));
-        if is_replica {
```

This would be a good refactoring, but not for the case where we need to manage replica-to-primary state transitions. The intended pattern is as follows: `PreferredBlobSender, CacheWarmUpExecutor, and PreferredSequencerDb` will handle the transition internally. To do this, they keep an inner: `inner: Option<_>`. When a replica becomes primary, `None` is updated to `Some(..)`. We can’t move this logic outside, as it would become too complicated. So let's move the flag and the Option back into all these tree components. (Soon the `Option` will become a 3 leg enum where we will also have `IsTransitioning` case)

---

Makes sense. I just don't like the idea of having objects that silently do nothing.
When reading code that does `self.blob_sender.store_blob()` I would expect it to actually store the blob.
In this case - it checks that it's a replica within and noops. In an ideal world - we would have `PrimarySequencer ` and `ReplicaSequencer` that would share some logic but also implement the differences and have smth like `replica.promote() -> Primary`, but agree that this is not the place to do this so reverted.